src/sys/socket/mod.rs

```diff
@@ -546,6 +550,20 @@ impl ControlMessageOwned {
                 let dl = ptr::read_unaligned(p as *const libc::in_addr);
                 ControlMessageOwned::Ipv4RecvDstAddr(dl)
             },
+            #[cfg(any(target_os = "android", target_os = "linux"))]
+            (libc::IPPROTO_IP, libc::IP_RECVERR) => {
+                let ee = p as *const libc::sock_extended_err;
+                let err = ptr::read_unaligned(ee);
+                let addr = ptr::read_unaligned(libc::SO_EE_OFFENDER(ee) as *const sockaddr_in);
+                ControlMessageOwned::Ipv4RecvErr(err, addr)
+            },
+            #[cfg(any(target_os = "android", target_os = "linux"))]
+            (libc::IPPROTO_IPV6, libc::IP_RECVERR) => {
+                let ee = p as *const libc::sock_extended_err;
+                let err = ptr::read_unaligned(ee);
+                let addr = ptr::read_unaligned(libc::SO_EE_OFFENDER(ee) as *const sockaddr_in6);
```

Comment:
`man 2 recvmsg`

> For local errors, no address is passed (this can be checked with the cmsg_len member  of  the  cmsghdr). 

Given that, I think `addr` should become an `Option`, and `len` should be handled appropriately.